contracts/src/ethereum/TokenBridge.sol

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: Apache-2.0.
+pragma solidity ^0.6.12;
+
+import "@joriksch/sg-contracts/src/starkware/contracts/components/GenericGovernance.sol";
+import "@joriksch/sg-contracts/src/starkware/contracts/interfaces/ContractInitializer.sol";
+import "@joriksch/sg-contracts/src/starkware/contracts/interfaces/ProxySupport.sol";
+import "@joriksch/sg-contracts/src/starkware/cairo/eth/CairoConstants.sol";
+import "@joriksch/sg-contracts/src/starkware/starknet/eth/StarknetMessaging.sol";
+
+interface IL1Token {
+    function approve(address, uint256) external returns (bool);
+    function balanceOf(address) external view returns (uint256);
+    function transferFrom(address, address, uint256) external returns (bool);    
+}
+
+contract TokenBridge is
+    GenericGovernance,
+    ContractInitializer,
+    ProxySupport
+{
+    event LogDeposit(address sender, address token, uint256 amount, uint256 l2Recipient);
+    event LogWithdrawal(address token, address recipient, uint256 amount);
+    event LogBridgeAdded(address l1Token, uint256 l2Token);
+
+    mapping(address => uint256) public l1TokentoL2Token;
+    StarknetMessaging public messagingContract;
+    uint256 l2TokenBridge;
+
+    constructor() public GenericGovernance("AAVE_BRIDGE_GOVERNANCE") {}
+
+    function both(bool x, bool y) internal pure returns (bool z) {
+        assembly{ z := and(x, y)}
+    }
+
+    function toSplitUint(uint256 value) internal pure returns (uint256, uint256) {
+      uint256 low = value & ((1 << 128) - 1);
+      uint256 high = value >> 128;
+      return (low, high);
+    }
+
+    function isInitialized() internal view override returns (bool) {
+        return messagingContract != StarknetMessaging(0);
+    }
+
+    function numOfSubContracts() internal pure override returns (uint256) {
+        return 0;
+    }
+
+    function validateInitData(bytes calldata data) internal pure override {
+        // 288 because address type (32) is replaced by uint256 type (256)
+        require(data.length == 288, "ILLEGAL_DATA_SIZE");
+    }
+
+    function processSubContractAddresses(bytes calldata subContractAddresses) internal override {}
+
+    modifier isApprovedToken(address token) {
+        uint256 l2TokenAddress = l1TokentoL2Token[token];
+        require(both(l2TokenAddress != 0, l2TokenAddress < CairoConstants.FIELD_PRIME), "L2_TOKEN_HAS_NOT_BEEN_APPROVED");
+        _;
+    }
+
+    /*
+      Gets the addresses of bridgedToken & messagingContract from the ProxySupport initialize(),
+      and sets the storage slot accordingly.
+    */
+    function initializeContractState(bytes calldata data) internal override {
```

Comment:
@swapnilraj @JorikSchellekens This is the function that is to be called when deploying the bridge behind a proxy.